File: golang/intensity-segments.go
package main

import (
	"fmt"
	"sort"
)

// IntensitySegments
//
//	Using a map (because of it dscreteness) to store: [segemnt] -> intensity
type IntensitySegments struct {
	it map[int]int
}

// NewIntensitySegments new a IntensitySegments Object
// return a pointer to the new object
func NewIntensitySegments() *IntensitySegments {
	return &IntensitySegments{
		it: map[int]int{},
	}
}

// set set a new range(from <-> to, to not included) of intensity based on existing one.
func (is *IntensitySegments) set(from, to int, amount int) {
	keys := is.orderedKeys()
	if len(keys) == 0 { // if no segments, in the intial case
		is.it[to] = 0
		is.it[from] = amount
		return
	}

	if to < keys[0] || to > keys[len(keys)-1] { // the end of set segment is outside of the existing one
		is.it[to] = 0
	} else if _, f := is.it[to]; !f { // 'to' is not found in existing segment numbers
		l := is.leftKey(to)
		is.it[to] = is.it[l]
	}

	is.it[from] = amount // directly set the segment beginning

	for _, k := range is.orderedKeys() { // all keys between from and to is invalid, delete them
		if k > from && k < to {
			delete(is.it, k)
		}
	}

	is.merge()
}

// add add a new range(from <-> to, to not included) of intensity to existing one.
func (is *IntensitySegments) add(from, to int, amount int) {
	keys := is.orderedKeys()
	if len(keys) == 0 { // if no segments, in the intial case
		is.set(from, to, amount)
		return
	}

	if to < keys[0] || from > keys[len(keys)-1] { // no overlap case
		is.set(from, to, amount)
		return
	}

	// handle the 'to'
	if _, f := is.it[to]; !f { // if not found to, add it
		l := is.leftKey(to)
		is.it[to] = is.it[l]
	}

	// handle the 'from'
	if from < keys[0] {
		is.it[from] = amount
	} else if from == keys[0] {
		is.it[from] += amount
	} else {
		if _, f := is.it[from]; !f {
			is.it[from] = amount
			if l := is.leftKey(from); l != -1 {
				is.it[from] += is.it[l]
			}
		} else {
			is.it[from] += amount
		}
	}

	// handle the keys between from and 'to'
	for _, k := range keys {
		if k > from && k < to {
			is.it[k] += amount
		}
	}

	// merge segments
	is.merge()
}

// merge merge continous segments to together if they have same intensity
func (is *IntensitySegments) merge() {
	keys := is.orderedKeys()
	if len(keys) == 0 {
		return
	}

	// clear prefix 0
	i := 0
	for ; i < len(keys) && is.it[keys[i]] == 0; i++ {
		delete(is.it, keys[i])
	}
	keys = keys[i:]

	for i = len(keys) - 1; i >= 0; i-- {
		if is.it[keys[i]] == 0 {
			if i-1 >= 0 && is.it[keys[i-1]] == 0 {
				delete(is.it, keys[i])
				keys = append(keys[0:i], keys[i+1:]...)
			}
		}

	}

	// clear the surffix segment with same intensity <> 0
	lastIntensity := is.it[keys[0]]
	for i := 1; i < len(keys); i++ {
		if is.it[keys[i]] == lastIntensity && lastIntensity != 0 {
			delete(is.it, keys[i])
		} else {
			lastIntensity = is.it[keys[i]]
		}
	}

	// clear the surffix segment with intensity == 0
	lastIntensity = 0
	for i := len(keys) - 2; i >= 0; i-- {
		if is.it[keys[i]] == 0 && lastIntensity == 0 {
			delete(is.it, keys[i])
		} else {
			lastIntensity = is.it[keys[i]]
		}
	}
}

// orderedKeys orderedKeys return a sorted keys of is.it
func (is *IntensitySegments) orderedKeys() []int {
	keys := intSlice{}
	for k := range is.it {
		keys = append(keys, k)
	}
	sort.Sort(keys)
	return keys
}

// dumps dumps return a string of simple format, i.e. [[20 1] [30 2] [40 0]]
func (is *IntensitySegments) dumps() string {
	keys := is.orderedKeys()
	rlt := [][2]int{}
	for _, k := range keys {
		rlt = append(rlt, [2]int{k, is.it[k]})
	}
	return fmt.Sprintf("%v", rlt)
}

// leftKey return the left segment number of the given.
func (is *IntensitySegments) leftKey(x int) int {
	l := -1
	for _, k := range is.orderedKeys() {
		if k < x {
			l = k
		}
	}
	return l
}

// toString print the dumped string simply
func (is *IntensitySegments) toString() {
	fmt.Printf("%v\n", is.dumps())
}

File: javascript/intensity-segments.js
/** *
 * NOTE: Feel free to add any extra member variables/functions you like.
 */
class IntensitySegments {

    /**
     * Using a map (because of it dscreteness) to store: [segemnt] -> intensity
     */
    constructor() {
        this.it = new Map()
    }
    /**
     * add a new range(from <-> to, to not included) of intensity to existing one.
     * @param {*} from 
     * @param {*} to 
     * @param {*} amount 
     * @returns 
     */
    add(from, to, amount) {
        // TODO: implement this
        let keys = this.orderedKeys()
        if (keys.length == 0) {
            this.set(from, to, amount)
            return
        }

        let lkeys = keys.length
        if (to < keys[0] || from > keys[lkeys - 1]) {
            this.set(from, to, amount)
            return
        }

        if (!this.it.has(to)) {
            let l = this.leftKey(to)
            this.it.set(to, this.it.get(l))
        }

        if (from < keys[0]) {
            this.it.set(from, amount)
        } else if (from == keys[0]) {
            this.it.set(from, this.it.get(from) + amount)
        } else {
            if (!this.it.has(from)) {
                this.it.set(from, amount)
                let l = this.leftKey(from)
                if (l != -1) {
                    this.it.set(from, this.it.get(from) + this.it.get(l))
                }
            } else {
                this.it.set(from, this.it.get(from) + amount)
            }
        }

        keys.forEach(k => {
            if (k > from && k < to) {
                this.it.set(k, this.it.get(k) + amount)
            }
        })

        this.merge()
    }
    /**
     * set a new range(from <-> to, to not included) of intensity based on existing one.
     * @param {*} from 
     * @param {*} to 
     * @param {*} amount 
     * @returns 
     */
    set(from, to, amount) {
        // TODO: implement this
        let keys = this.orderedKeys()
        if (keys.length == 0) {
            this.it.set(to, 0)
            this.it.set(from, amount)
            return
        }

        let lkeys = keys.length
        if (to < keys[0] || to > keys[lkeys - 1]) {
            this.it.set(to, 0)
        } else if (!this.it.has(to)) {
            let l = this.leftKey(to)
            this.it.set(to, this.it.get(l))
        }

        this.it.set(from, amount)

        keys = this.orderedKeys()
        for (k in keys) {
            if (k > from && k < to) {
                this.it.delete(k)
            }
        }

        this.merge()
    }
    /**
     * print the dumped string simply
     */
    toString() {
        // TODO: implement this
        console.log(this.dumps())
    }

    /**
     * get ordered keys of 'it' member.
     * @returns a sorted keys of is.it
     */
    orderedKeys() {
        let keys = this.it.keys()
        let ks = Array.from(keys)
        ks.sort()
        return ks
    }

    /**
     * get the left segment number for x from this.it
     * @param {*} x 
     * @returns the left segment number of the given.
     */
    leftKey(x) {
        let l = -1
        let keys = this.orderedKeys()
        keys.forEach(k => {
            if (k < x) {
                l = k
            }
        })
        return l
    }

    /**
     * merge continous segments to together if they have same intensity
     * @returns 
     */
    merge() {
        let keys = this.orderedKeys()
        if (keys.length == 0) {
            return
        }

        let i = 0
        for (; i < keys.length && this.it.get(keys[i]) == 0; i++) {
            this.it.delete(keys[i])
        }
        keys.splice(0, i)

        for (i = keys.length - 1; i >= 0; i--) {
            if (this.it.get(keys[i]) == 0) {
                if (i - 1 >= 0 && this.it.get(keys[i - 1]) == 0) {
                    this.it.delete(keys[i])
                    keys.splice(i, 1)
                }
            }
        }

        let last
        last = this.it.get(keys[0])
        for (let i = 1; i < keys.length; i++) {
            if (this.it.get(keys[i]) == last && last != 0) {
                this.it.delete(keys[i])
            } else {
                last = this.it.get(keys[i])
            }
        }

        last = 0
        for (let i = keys.length - 2; i >= 0; i--) {
            if (this.it.get(keys[i]) == 0 && last == 0) {
                this.it.delete(keys[i])
            } else {
                last = this.it.get(keys[i])
            }
        }
    }

    /**
     * dumps return a string of simple format, i.e. [[20 1] [30 2] [40 0]]
     * @returns 
     */
    dumps() {
        let keys = this.orderedKeys()
        let rlt = []
        keys.forEach(k => {
            rlt.push(`[${k},${this.it.get(k)}]`)
        })
        return `[${rlt.join(",")}]`
    }
}

module.exports = IntensitySegments